core/src/main/scala/org/broadinstitute/dsde/rawls/fastpass/FastPassService.scala

```diff
@@ -0,0 +1,444 @@
+  ): Future[Unit] = {
+    logger.info(
+      s"Adding project-level FastPass access for $userAndPet in ${googleProjectId.value} [${organizationRoles.mkString(" ")}]"
+    )
+    for {
+      _ <- googleIamDao.addRoles(
+        GoogleProject(googleProjectId.value),
+        userAndPet.userEmail,
+        IamMemberTypes.User,
+        organizationRoles,
+        condition = Some(condition)
+      )
+      _ <- openTelemetry.incrementCounter("fastpass-iam-granted-user", tags = openTelemetryTags).unsafeToFuture()
+      _ <- googleIamDao.addRoles(
+        GoogleProject(googleProjectId.value),
+        userAndPet.petEmail,
+        IamMemberTypes.ServiceAccount,
+        organizationRoles,
+        condition = Some(condition)
+      )
+      _ <- openTelemetry.incrementCounter("fastpass-iam-granted-pet", tags = openTelemetryTags).unsafeToFuture()
+    } yield ()
+  }
+
+  private def addUserAndPetToBucketIamRole(gcsBucketName: GcsBucketName,
+                                           organizationRoles: Set[String],
+                                           userAndPet: UserAndPetEmails,
+                                           condition: Expr,
+                                           googleProjectId: GoogleProjectId
+  ): Future[Unit] = {
+    logger.info(
+      s"Adding bucket-level FastPass access for $userAndPet in ${gcsBucketName.value} [${organizationRoles.mkString(" ")}]"
+    )
+    for {
+      _ <- googleStorageDAO.addIamRoles(
+        gcsBucketName,
+        userAndPet.userEmail,
+        IamMemberTypes.User,
+        organizationRoles,
+        condition = Some(condition),
+        userProject = Some(GoogleProject(googleProjectId.value))
+      )
+      _ <- openTelemetry.incrementCounter("fastpass-iam-granted-user", tags = openTelemetryTags).unsafeToFuture()
+      _ <- googleStorageDAO.addIamRoles(
+        gcsBucketName,
+        userAndPet.petEmail,
+        IamMemberTypes.ServiceAccount,
+        organizationRoles,
+        condition = Some(condition),
+        userProject = Some(GoogleProject(googleProjectId.value))
+      )
+      _ <- openTelemetry.incrementCounter("fastpass-iam-granted-pet", tags = openTelemetryTags).unsafeToFuture()
+    } yield ()
+  }
+
+  private def quotaAvailableForNewWorkspaceFastPassGrants(workspace: Workspace): Future[Boolean] =
+    for {
+      projectPolicy <- googleIamDao
+        .getProjectPolicy(GoogleProject(workspace.googleProjectId.value))
+        .map(fromProjectPolicy)
+      bucketPolicy <- googleStorageDAO
+        .getBucketPolicy(GcsBucketName(workspace.bucketName), Some(GoogleProject(workspace.googleProjectId.value)))
+        .map(fromBucketPolicy)
+    } yield {
+      // Role binding quotas do not de-duplicate member emails, hence the conversion of Sets to Lists
+      val numProjectRoleBindings = projectPolicy.bindings.toList.flatMap(_.members.toList).size
+      val expectedProjectBindings = numProjectRoleBindings + maxProjectRoleBindingsPerUser
```

Comment:
Why is the max number being added on?

Reply:
added a clarifying comment, this function is to check whether we can complete the operation and stay below the quota

Reply:
But isn't `numProjectRoleBindings` the actual number of bindings that will be added?

Reply:
i think the confusion is coming from the variable names `maxProjectRoleBindingsPerUser` this is the number of role bindings we will be adding in the operation not the total quota. so numProjectRoleBindings + maxProjectRoleBindingsPerUser should be less than the quota to go forward with the operation